### PR TITLE
Add ability to exclude paths when searching files

### PR DESCRIPTION
**Pull request by Seidr**
I started looking into this when a colleague mentioned the lack of exclusion when searching files in Atom. I'm not sure whether or not what I've done is the _correct_ way of doing it, but from what I can see it is working.

Any search paths that a prefixed with either the tilde or exclamation mark character will be excluded from the scan, with the excluded paths as a prefix (although it appears to work using the glob syntax, as well).

Referencing issue atom/find-and-replace#149

### CODE REVIEW

File: spec/workspace-spec.coffee [hunk 1]
```diff
@@ -878,6 +878,24 @@ describe "Workspace", ->
           runs ->
             expect(resultPaths.sort()).toEqual([file1, file2].sort())
 
+        describe "when an inclusion path starts with either a tilde or exclamation mark", ->
+          it "the path is used as part of the exclusion array, instead of inclusion", ->
```
Comment (50Wliu) [inline spec/workspace-spec.coffee]: >excludes that path instead of including it

or just
>excludes that path

Comment (Seidr) [inline spec/workspace-spec.coffee]: Agreed - it did feel like a bit of a ramble by the time I finished writing it. I'll update it.

File: src/workspace.coffee [hunk 2]
```diff
@@ -792,10 +792,10 @@ class Workspace extends Model
 
     searchOptions =
       ignoreCase: regex.ignoreCase
-      inclusions: options.paths
+      inclusions: (path for path in options.paths when path[0]!="!" && path[0]!="~")
```
Comment (mnquintana) [inline src/workspace.coffee]: Use `and` instead of `&&`

Comment (mnquintana) [inline src/workspace.coffee]: And `isnt` instead of `!=`

File: src/workspace.coffee [hunk 3]
```diff
@@ -792,10 +792,10 @@ class Workspace extends Model
...
       includeHidden: true
       excludeVcsIgnores: atom.config.get('core.excludeVcsIgnoredPaths')
-      exclusions: atom.config.get('core.ignoredNames')
+      exclusions: atom.config.get('core.ignoredNames').concat(path.substring(1) for path in options.paths when path[0]=="!" || path[0]=="~")
```
Comment (mnquintana) [inline src/workspace.coffee]: Use `is` instead of `==`, and `or` instead of `||`

**Comment by Seidr** [comment]
Damn - not sure I meant to push that merge, sorry. Just making the changes you pointed out @mnquintana

**Comment by Seidr** [comment]
There we go - will that merge cause an issue for this pull request? I'm used to getting my local copy up to date prior to pushing anything (coming from primarily a Mercurial background..)

**Comment by Seidr** [comment]
Thanks for the feedback @mnquintana - I've made the changes you've requested to bring this pull request in line with the styleguides.

**Comment by benogle** [comment]
Thanks for this. It might take us a bit to get to a deeper review.

**Comment by saitodisse** [comment]
This is important, guys. :+1:

**Comment by cardoni** [comment]
:+1:

**Comment by thadk** [comment]
:+1:

**Comment by dbkaplun** [comment]
It's been two months... is that long enough to review a 2-line change?

**Comment by madhancock** [comment]
+1

**Comment by Seidr** [comment]
@benogle Just wondering if there's been any progress with this review? Thanks

**Comment by benogle** [comment]
Sorry about the delay. This has stalled because we werent (arent) sure if this was the way we wanted to implement this. Ideally, the negation handling would take place in [scandal](github.com/atom/scandal), and come from minimatch, which allegedly supports [negation](https://github.com/isaacs/minimatch#properties). One of us (likely me) was going to dig in, but we've been focused on other things.

Realistically, this wont happen for at least 2 weeks due to our current focus on the 1.0 launch.

Thanks for your patience, and for writing specs.

**Comment by Seidr** [comment]
Thanks for the feedback, and no problem re further delays. Completely understand your concerns RE thos change. I look forward to 1.0, and having this feature included at some point in the future :)

**Comment by thedaniel** [comment]
Looks like this needs to be handled in [scandal](https://github.com/atom/scandal) as Ben notes above. Closing - thanks for getting the conversation started on negation.

**Comment by thedaniel** [comment]
Looks like this needs to be handled in [scandal](https://github.com/atom/scandal) as Ben notes above. Closing - thanks for getting the conversation started on negation.

**Comment by benogle** [comment]
My plan is to start work on this very soon as it is now part of our focus: https://github.com/atom/atom/issues/7995

**Comment by benogle** [comment]
A fix to this is out in atom 1.0.3. See atom/scandal#29 for details

**Pull request closed without merging**